### PR TITLE
Recipe app

### PR DESCRIPTION
**Pull request by MohamedHNoor**
# Hi Code Reviewer 👋 👋 

## In this pull request, the following changes have been made:

- [x]  A login page using devise.
- [x] A registration page using devise.
- [x] A food list.
- [x] A recipe list.
- [x] A general shopping list view.
- [x] A list of all public recipes from other users with their names and total prices.
- [x] Add rspec tests
- [x] Fix n+1 problems

### CODE REVIEW

**Review by FrederickMih — CHANGES_REQUESTED** [review]
## Hi :raising_hand:  @MohamedHNoor @shegy28 , 


Good job so far!
There are some issues that you still need to work on to prepare your project for the final evaluation but you are almost there!


## Highlights :green_circle: 

- PR is professional :+1: 
- Design of app matches required wireframes :rocket: :rocket: 
- Authentication system has been correctly set up :+1: 
- All unit tests run without errors :heavy_check_mark: 
- Descriptive README.md file :+1: 
- Great commit history :+1: 
- Linter checks passed :heavy_check_mark: 

## Optional suggestions
**N/A**

You can use as many of my suggestions as you want. If there is anything you would like to skip - feel free to do that. However, I strongly recommend you to take them into account as they can make your code better._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

File: app/views/foods/index.html.erb [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+      <th>Quantity</th>
+      <th>Action</th>
+    </tr>
+  </thead>
+  <tbody>
+    <% @foods.each do |food| %>
+      <tr class="text-center fw-bold">
+        <td><%= food.name %></td>
+        <td><%= food.measurement_unit %></td>
+        <td><%= food.price %></td>
```
[the remark this answers is not in the document]
Comment (MohamedHNoor) [reply app/views/foods/index.html.erb]: Done

File: app/views/foods/index.html.erb [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+        <td><%= food.quantity %></td>
```
[the remark this answers is not in the document]
Comment (MohamedHNoor) [reply app/views/foods/index.html.erb:22]: Done

File: app/views/recipes/show.html.erb [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+    </div>
+  </div>
+  <hr>
+  <div class="container">
+    <table class="table table-bordered table-striped">
+      <thead>
+        <tr class="text-center">
+          <th>Food</th>
+          <th>Quantity</th>
+          <th>Value</th>
```
[the remark this answers is not in the document]
Comment (MohamedHNoor) [reply app/views/recipes/show.html.erb:44]: Done

File: app/views/recipes/shopping_list.html.erb [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+       </div>
+    </div>
+ </div>
+ <div class="container">
+    <table class="table table-bordered table-striped">
+      <thead>
+        <tr class="text-center">
+          <th>Food</th>
+          <th>Quantity</th>
+          <th>Price</th>
```
[the remark this answers is not in the document]
Comment (MohamedHNoor) [reply app/views/recipes/shopping_list.html.erb]: Done

File: app/views/recipes/show.html.erb [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+          <%= form_with(url: user_recipe_path(@recipe)) do |form| %>
+            Public: <%= form.check_box :public, { checked: @recipe.public, disabled: true }, class: 'text-white bg-dark public' %>
+          <% end %>
+        </div>
+      </div>
+    </div>
+
+    <div class="row">
+      <div class="d-flex justify-content-between pb-4">
+        <%= link_to 'Generate Shopping List', shopping_list_path(@recipe.id), class: "btn btn-outline-secondary"  %>
```
[the remark this answers is not in the document]
Comment (MohamedHNoor) [reply app/views/recipes/show.html.erb]: Done

File: config/routes.rb [hunk 6]
```diff
@@ -0,0 +1,23 @@
+Rails.application.routes.draw do
+  devise_for :users
+  # Define your application routes per the DSL in https://guides.rubyonrails.org/routing.html
+
+  # Defines the root path route ("/")
+  root to: "foods#index"
+
+  # get 'recipe_foods/:recipe_id', to: 'recipe_foods#new', as: 'new_recipe_food'
+  # post 'recipe_foods/:recipe_id', to: 'recipe_foods#create', as: 'create_recipe_food'
```
[the remark this answers is not in the document]
Comment (MohamedHNoor) [reply config/routes.rb]: Done

File: config/routes.rb [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+  
+  resources :users, only: [:index, :show]  do
+    resources :foods
+    resources :recipes do
+      resources :recipe_foods
+    end
+  end
+
+  # post '/users/:user_id/recipes/:recipe_id/recipe_foods', to: 'recipe_foods#create', as: 'create_recipe_recipe_food'
```
[the remark this answers is not in the document]
Comment (MohamedHNoor) [reply config/routes.rb]: Done

File: config/routes.rb [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+  # Defines the root path route ("/")
+  root to: "foods#index"
+
+  # get 'recipe_foods/:recipe_id', to: 'recipe_foods#new', as: 'new_recipe_food'
+  # post 'recipe_foods/:recipe_id', to: 'recipe_foods#create', as: 'create_recipe_food'
+  
+  resources :users, only: [:index, :show]  do
+    resources :foods
+    resources :recipes do
+      resources :recipe_foods
```
[the remark this answers is not in the document]
Comment (MohamedHNoor) [reply config/routes.rb]: Done

File: app/models/user.rb [hunk 9]
```diff
@@ -0,0 +1,10 @@
+class User < ApplicationRecord
+  # Include default devise modules. Others available are:
+  # :confirmable, :lockable, :timeoutable, :trackable and :omniauthable
```
[the remark this answers is not in the document]
Comment (MohamedHNoor) [reply app/models/user.rb]: Done

File: app/views/recipes/_public_recipes.html.erb [hunk 10]
```diff
@@ -0,0 +1,14 @@
...
+  <div class="d-flex align-items-center justify-content-between p-3">
+    <div>
+      <h2><%= link_to user_recipe_path(recipe.user, recipe), class: "text-decoration-none text-dark" do %>
+          <%= recipe.name%>
+        <% end %></h2>
+      <p>By: <%= recipe.user.name %></p>
+    </div>
+    <div>
+      <p>Total Food Items: <%=recipe.recipe_foods.length%></p>
+      <p>Total Price: $<%= @total_price[recipe.id] %></p>
```
[the remark this answers is not in the document]
Comment (MohamedHNoor) [reply app/views/recipes/_public_recipes.html.erb]: Done

**Review by KayLemba — APPROVED** [review]
[[# STATUS: APPROVED :green_circle: :clap: 

Hello @TEAM        , 
Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉  Keep up the good work 🥂 

HIGHLIGHTS:

- [x]  Your repo is professional  😎 
- [x] Descriptive P.R 💯 
- [x] linters are passing
- [x] Professional README 💯
- [x] Tests are passing 
- [x] Nice UI design 🥇 
- [ ] Receipt App fully Functional 💯 🥇 🥂 
 










 








Now you can merge this branch and move on :1st_place_medal: 
Happy Coding cheers 🥇 💯 



well-1

Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better.

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._]]

**Pull request merged**